### PR TITLE
Parallel-netcdf: update package.

### PR DESCRIPTION
**Pull request by skosukhin**
- optionally (and by default) enable shared libraries
- apply a patch for NAG
- refactoring

### CODE REVIEW

File: var/spack/repos/builtin/packages/parallel-netcdf/package.py [hunk 1]
```diff
@@ -25,8 +25,7 @@ def url_for_version(self, version):
 
         return url.format(version.dotted)
 
-    version('develop', branch='develop')
-    version('master', branch='master')
+    version('develop', branch='master')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/parallel-netcdf/package.py]: Why these changes?

File: var/spack/repos/builtin/packages/parallel-netcdf/package.py [hunk 2]
```diff
@@ -42,41 +41,64 @@ def url_for_version(self, version):
...
     depends_on('m4', type='build')
 
-    # See:
-    # https://trac.mcs.anl.gov/projects/parallel-netcdf/browser/trunk/INSTALL
-    def configure_args(self):
-        spec = self.spec
+    conflicts('+shared', when='@:1.9%nag+fortran')
+    conflicts('+shared', when='@:1.8')
+
+    patch('nag_libtool.patch', when='@1.10:999%nag')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/parallel-netcdf/package.py]: I take it this has been fixed upstream? In that case, you might want to do something like:
```suggestion
    patch('nag_libtool.patch', when='@1.10:1.12.1%nag')
```

File: var/spack/repos/builtin/packages/parallel-netcdf/package.py (hunk 1 above)
Comment (skosukhin) [reply var/spack/repos/builtin/packages/parallel-netcdf/package.py]: There is no `develop` branch in the repository actually. Only `master` and, as far as I remember, version `develop` has a special meaning in Spack and `master` does not. So, I made branch `master` the `develop` version.

File: var/spack/repos/builtin/packages/parallel-netcdf/package.py (hunk 2 above)
Comment (skosukhin) [reply var/spack/repos/builtin/packages/parallel-netcdf/package.py]: This has not been fixed upstream and hardly will be in the nearest future because it would require a special `autogen` script, which would apply patches to the libtool-generated files. Developers are usually very reluctant to switch from a simple `autoreconf -fvi` to something more complex with an `autogen` script and a bunch of patches for libtool. The reason I introduced the `999` constraint was not to apply the patch to the `develop` version, which does not have files that are modified with the patch.

**Comment by skosukhin** [comment]
I could become a maintainer just to be notified and review future changes. However, I am not that interested in this package to keep track of new version and add them to Spack.

File: var/spack/repos/builtin/packages/parallel-netcdf/package.py (hunk 2 above)
Comment (skosukhin) [reply var/spack/repos/builtin/packages/parallel-netcdf/package.py]: It turns out that I was wrong: there is already a patched version of `libtool.m4` upstream so no `autogen.sh` is really needed. The patch for `ltmain.sh` can be handled the same way: https://github.com/Parallel-NetCDF/PnetCDF/pull/59

File: var/spack/repos/builtin/packages/parallel-netcdf/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/parallel-netcdf/package.py]: See #1983, both master and develop have special handling, so you can call this master

**Comment by skosukhin** [comment]
@adamjstewart yes, it could be much shorter but that would require patching after `autoreconf` stage or as part of it. Instead, I suggest the we drop the patch for `master` and I will try to get this fixed upstream.
